### DISCUSSION TITLE
Is there any way to query from a zipped(*.vcf.gz) and indexed(*.vcf.gx.tbi) file with chr,position,ref and alt("chr2:9452897_AT_T)")

### DISCUSSION THREAD

**Discussion by amirul99** [samtools/htsjdk#1627] [Q&A] [answered]
Hi, 
I'm using TabixReader to retrive data from a large vcf zipped file. I need a specific output row with mathced chr,position,ref and alt.

**Comment by lindenb** [comment, marked as answer]
use a VCFFileReader + query https://samtools.github.io/htsjdk/javadoc/htsjdk/htsjdk/variant/vcf/VCFFileReader.html#query-java.lang.String-int-int- , and loop over the iterator/Variant

 > I need a specific output row with mathced chr,position,ref and alt.

this is then just basic programming...

**Reply by amirul99** [reply]
@lindenb thanks. TabixReader query() and VCFFileReader query() both are same maybe, I need to minimize the query time for my variants. That's why I'm trying to find a way to query with chr,pos,ref and alt.

**Reply by lbergelson** [reply]
@amirul99 I'm not sure I understand what you want exactly.  Querying by chrom/pos is fairly efficient.  There's no faster API to filter it by the alt alleles, and I'm not sure that it is possible to implement something faster than using the index to find relevant positions and then checking the alt contigs for a match.

What is your data / access pattern like?

**Reply by amirul99** [reply]
Hi @lbergelson, I'm working with gnomAd3 version, which has whole genome data. I'm trying to retrieve data for my variants from their zipped and indexed file. It tooks time to retrieve my data. That's why I'm wondering how faster can I retrieve data with htsjdk. It will helpful for me.

**Reply by lindenb** [reply]
@amirul99 when I'm scanning for a resource like gnomad with many variants, I use a 'buffered' VCF reader where I load all the variants in a large region. https://github.com/lindenb/jvarkit/blob/master/src/main/java/com/github/lindenb/jvarkit/variant/vcf/BufferedVCFReader.java

**Reply by lbergelson** [reply]
@lindenb That can help a ton if your queries are close enough together, if they're very sparse though it could  potentially slow you down since gnomad is so dense.  GATK has a similar idea with its FeatureDataSource that does caching.

@amirul99  Gnomad is just inherently very large though and htsjdk's vcf reading is not as well optimized as it could ideally be.  If you're using one of the files that includes genotypes you could first scan the site-only version in order to identify sites that match your allele filters faster.
